### PR TITLE
[1651] Apply review draft page

### PR DESCRIPTION
**Pull request by JonF1**
### Context
https://trello.com/c/0SrjkmaA/1651-l-apply-review-draft-page

### Changes proposed in this pull request
* Apply draft and draft component wrapped in component that decides which one to render depending on whether trainee has an Apply application
* Review draft show view refactored to use this wrapper component
* Task list helper to keep component views DRY
* Page title translations dependent on trainee type
* Show view spec and component specs
* Review status out of scope

### Guidance to review
Create trainee in rails console to test apply version of review page in browser 
Create apply application with required fields
Associate trainee with application
Toggle `training_route` field on trainee to check provider-led, school direct etc

### CODE REVIEW

File: app/helpers/task_list_helper.rb [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      return if task_list[:training_details][:task_name] != "Training details"
+
+      task_list[:training_details][:task_name] = "Trainee start date and ID"
+    end
+
+    task_list[task]
+  end
+end
+
+def draft_only_row_helper(trainee, task)
```
Comment (AbigailMcP) [inline app/helpers/task_list_helper.rb:78]: Why does this config need to be in a separate helper method? Can these items not live in the `task_list` in the general `row_helper` above?

File: app/helpers/task_list_helper.rb [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+          confirm_path: edit_trainee_lead_schools_path(trainee),
+          classes: "school-details",
+          status: ProgressService.call(
+            validator: Schools::FormValidator.new(trainee),
+            progress_value: trainee.progress.schools,
+          ).status,
+        },
+    }
+
+    unless trainee.apply_application?
```
Comment (AbigailMcP) [inline app/helpers/task_list_helper.rb:68]: This is the switcheroonie code.

File: app/components/review_draft/draft/view.html.erb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+<%= render Trainees::RouteIndicator::View.new(trainee: @trainee) %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+
+    <h2 class="govuk-heading-m">Personal details and education</h2>
+
+    <%= render TaskList::View.new do |component|
+          component.row(**draft_only_row_helper(@trainee, :personal_details))
```
Comment (AbigailMcP) [inline app/components/review_draft/draft/view.html.erb:9]: Indentation.

File: app/components/review_draft/application_draft/view.html.erb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+              )
+          end
+
+          if show_publish_courses?(@trainee)
+            component.row(**row_helper(@trainee, :publish_course_details))
+          else
+            component.row(**row_helper(@trainee, :course_details))
+          end
+
+          component.row(
```
Comment (AbigailMcP) [inline app/components/review_draft/application_draft/view.html.erb:27]: Can this go in your new `row_helper` method?

File: app/views/trainees/review_draft/show.html.erb [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<%= render PageTitle::View.new(title: "trainees.show", has_errors: @trainee.errors.present?) %>
+<%= render PageTitle::View.new(title: "trainees.show.apply_draft", has_errors: @trainee.errors.present?) %>
```
[the remark this answers is not in the document]
Comment (JonF1) [reply app/views/trainees/review_draft/show.html.erb]: fixed

File: app/views/trainees/review_draft/show.html.erb [hunk 6]
```diff
@@ -8,133 +8,11 @@
...
 
 <h1 class="govuk-heading-l">
-  <% if @trainee.first_names.present? && @trainee.last_name.present? %>
-    <span class="govuk-caption-l">Draft record for <%= trainee_name(@trainee) %></span>
-  <% else %>
-    <span class="govuk-caption-l">Draft record</span>
-  <% end %>
-  Add a trainee
+  <span class="govuk-caption-l"><%= trainee_draft_title(@trainee) %></span>
+  <%= t("components.page_titles.trainees.show.#{@trainee.apply_application? ? 'apply_draft' : 'draft'}") %>
```
[the remark this answers is not in the document]
Comment (JonF1) [reply app/views/trainees/review_draft/show.html.erb]: I've added to component

File: app/components/review_draft/application_draft/view.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module ReviewDraft
+  module ApplicationDraft
```
[the remark this answers is not in the document]
Comment (JonF1) [reply app/components/review_draft/application_draft/view.rb]: Applydraft is so much better, i've changed

File: app/components/review_draft/application_draft/view.html.erb [hunk 8]
```diff
@@ -0,0 +1,49 @@
+<%= render Trainees::RouteIndicator::View.new(trainee: @trainee) %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds-from-desktop">
+
+    <h2 class="govuk-heading-m">Registration data from Apply</h2>
```
[the remark this answers is not in the document]
Comment (JonF1) [reply app/components/review_draft/application_draft/view.html.erb]: thanks, fixed

File: app/components/review_draft/application_draft/view.html.erb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    <p class="govuk-body"> Application submitted on 8 March 2021, and recruited in Manage on 2 May 2021 </p>
```
[the remark this answers is not in the document]
Comment (JonF1) [reply app/components/review_draft/application_draft/view.html.erb]: I've removed, and I'll make a ticket - should i put in current sprint or ideas board it

File: app/components/review_draft/application_draft/view.html.erb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    <%= render TaskList::View.new do |component|
+          unless @trainee.requires_placement_details?
+            component.row(
+                task_name: "Review the trainee's application",
```
[the remark this answers is not in the document]
Comment (JonF1) [reply app/components/review_draft/application_draft/view.html.erb]: removed, must have been changed on prototype midway through 🤔

File: app/components/review_draft/view.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+module ReviewDraft
+  class View < GovukComponent::Base
+    def initialize(trainee:)
+      @trainee = trainee
+    end
+
+    def component
+      if @trainee.apply_application?
+        ReviewDraft::ApplicationDraft::View.new(trainee: @trainee)
```
[the remark this answers is not in the document]
Comment (JonF1) [reply app/components/review_draft/view.rb]: cheers, fixed

File: app/helpers/task_list_helper.rb (hunk 1 above)
Comment (JonF1) [reply app/helpers/task_list_helper.rb]: I've put them all into the row helper method 👍

File: app/helpers/task_list_helper.rb [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+          classes: "course-details",
+          status: ProgressService.call(
+            validator: CourseDetailsForm.new(trainee),
+            progress_value: trainee.progress.course_details,
+          ).status,
+        },
+
+      training_details:
+        {
+          task_name: "Training details",
```
[the remark this answers is not in the document]
Comment (JonF1) [reply app/helpers/task_list_helper.rb]: my new favourite phrase

File: app/components/review_draft/application_draft/view.html.erb (hunk 9 above)
[the remark this answers is not in the document]
Comment (AbigailMcP) [reply app/components/review_draft/application_draft/view.html.erb]: Ideas board please! I don't think we'll get round to it this sprint.

File: spec/components/review_draft/draft/view_spec.rb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+require "rails_helper"
+
+RSpec.describe ReviewDraft::Draft::View do
+  alias_method :component, :page
+
+  before do
+    render_inline(described_class.new(trainee: trainee))
+  end
+
+  context "sections that appear for all trainees regardless of route" do
```
Comment (AbigailMcP) [inline spec/components/review_draft/draft/view_spec.rb:12]: This context isn't _quite_ correct if you're doing `let(:trainee) { create(:trainee) }` - our base trainee factory actually always creates an Assessment only trainee. So this is only testing sections that appear for AO trainees, not 'regardless of route'.

File: spec/components/review_draft/draft/view_spec.rb [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    let(:trainee) { create(:trainee) }
+
+    it "shows for all trainees" do
```
Comment (AbigailMcP) [inline spec/components/review_draft/draft/view_spec.rb:15]: Please could we also test that some sections _don't_ appear, so we're guarding against the bug that graham was speaking about this morning i.e. schools section, placement section

File: spec/components/review_draft/draft/view_spec.rb [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+    it "shows provider-led specific sections" do
+      expect(component).to have_text("Placement details")
+    end
+  end
+
+  context "when the trainee is on the school direct routes", 'feature_routes.school_direct_salaried': true do
+    let(:trainee) { create(:trainee, :school_direct_salaried) }
+
+    it "shows provider-led specific sections" do
```
Comment (AbigailMcP) [inline spec/components/review_draft/draft/view_spec.rb:36]:
```suggestion
    it "shows school direct specific sections" do
```

File: spec/components/review_draft/draft/view_spec.rb [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+  context "when the trainee is on the provider-led route", 'feature_routes.provider_led_postgrad': true do
+    let(:trainee) { create(:trainee, :provider_led_postgrad, :with_placement_assignment) }
+
+    it "shows provider-led specific sections" do
+      expect(component).to have_text("Placement details")
+    end
+  end
+
+  context "when the trainee is on the school direct routes", 'feature_routes.school_direct_salaried': true do
```
Comment (AbigailMcP) [inline spec/components/review_draft/draft/view_spec.rb:33]: Could we test for both school direct routes here?

something like:

```ruby
%i[school_direct_salaried school_direct_tuition_fee].each do |route|
  context "when the trainee is on the #{route} routes", 'feature_routes.#{route}': true do
    let(:trainee) { create(:trainee, route) }

    it 'blah'
      'test'
    end
  end
end
```

File: spec/components/review_draft/apply_draft/view_spec.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+  describe "sections that appear for application draft trainee" do
+    before do
+      render_inline(described_class.new(trainee: trainee))
+    end
+
+    context "when the trainee is on the provider-led route", 'feature_routes.provider_led_postgrad': true do
+      let(:trainee) { create(:trainee, :provider_led_postgrad, :with_placement_assignment, :with_apply_application) }
+
+      it "renders the correct sections" do
```
Comment (AbigailMcP) [inline spec/components/review_draft/apply_draft/view_spec.rb:16]: As above, are there sections that specifically _shouldn't_ render? If so, test here.

File: spec/components/review_draft/apply_draft/view_spec.rb [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        expect(component).to have_text("Placement details")
+        expect(component).to have_text("Training details")
+      end
+    end
+
+    context "when the trainee is on the school direct route", 'feature_routes.school_direct_salaried': true do
+      let(:trainee) { create(:trainee, :school_direct_salaried, :with_apply_application) }
+
+      it "renders the correct sections" do
+        expect(component).to have_text("Course details")
```
Comment (AbigailMcP) [inline spec/components/review_draft/apply_draft/view_spec.rb:28]: As above.

File: spec/components/review_draft/apply_draft/view_spec.rb [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        expect(component).to have_text("Trainee data")
+        expect(component).to have_text("Placement details")
+        expect(component).to have_text("Training details")
+      end
+    end
+
+    context "when the trainee is on the school direct route", 'feature_routes.school_direct_salaried': true do
```
Comment (AbigailMcP) [inline spec/components/review_draft/apply_draft/view_spec.rb:24]: We should also make sure it renders correctly for `school_direct_tuition_fee` trainees - see code snippet for other spec.

File: spec/components/review_draft/draft/view_spec.rb (hunk 13 above)
Comment (JonF1) [reply spec/components/review_draft/draft/view_spec.rb]: Makes sense! i've changed to assessment only

File: spec/components/review_draft/draft/view_spec.rb (hunk 14 above)
Comment (JonF1) [reply spec/components/review_draft/draft/view_spec.rb]: i've added extra tests as you said to guard against the wrong sections rendering for wrong routes

File: spec/components/review_draft/draft/view_spec.rb (hunk 16 above)
Comment (JonF1) [reply spec/components/review_draft/draft/view_spec.rb]: Thanks, I've added

File: spec/components/review_draft/apply_draft/view_spec.rb (hunk 19 above)
Comment (JonF1) [reply spec/components/review_draft/apply_draft/view_spec.rb]: cheers

File: spec/components/review_draft/apply_draft/view_spec.rb (hunk 17 above)
Comment (JonF1) [reply spec/components/review_draft/apply_draft/view_spec.rb]: yep, I've added some, thanks

File: app/components/review_draft/application_draft/view.html.erb (hunk 9 above)
[the remark this answers is not in the document]
Comment (JonF1) [reply app/components/review_draft/application_draft/view.html.erb]: i've added it under my column 👍

**Review by AbigailMcP — APPROVED** [review]
👍 good for product review!

I noticed a bug that's definitely not to do with this ticket, so I'll card up separately - it's more just a heads up for the person doing the product review. We're turning _all_ in progress sections into pink 'REVIEW' ones for apply drafts. I believe it should only be the sections in **Registration data from Apply**:

Screenshot 2021-05-12 at 16 04 23

File: app/helpers/task_list_helper.rb [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+            validator: Schools::FormValidator.new(trainee),
+            progress_value: trainee.progress.schools,
+          ).status,
+        },
+
+      trainee_data:
+        {
+          task_name: "Trainee data",
+          path: "#",
+          confirm_path: "#",
```
Comment (FriedSock) [inline app/helpers/task_list_helper.rb:71]: The trainee data page has been added now https://github.com/DFE-Digital/register-trainee-teachers/pull/864 so we can add the link.

Comment (JonF1) [reply app/helpers/task_list_helper.rb]: I've added the path, though I'm not sure how the `progress_value` would work here as it's a combination of different sections being completed, probs beyond scope of ticket?

**Comment by GeoffJohnson** [comment]
All good

**Pull request merged**